### ISSUE TITLE
Training MatterGen on Custom Crystal Structures for Bandgap

### ISSUE THREAD

**Issue by msehabibur** [microsoft/mattergen#128]
I want to train the MatterGen model on my own dataset to generate new crystal structures based on targeted bandgap values. To prepare the data, I created a CSV file where each row contains the name of the compound or folder, the full content of the CIF file (not just a path), and the corresponding DFT-computed bandgap value . My goal is for the model to learn structure–property relationships and generate novel materials with desired electronic properties. My question is: is this data format sufficient for training MatterGen effectively, or are there additional preprocessing steps or formatting requirements needed to ensure the model can learn and generate materials conditioned on the bandgap? Also, when I have my custom dataset, how do I write the command to train the model? `mattergen-train data_module=mp_20 ~trainer.logger`

<img width="480" alt="Image" src="https://github.com/user-attachments/assets/2297bf9f-1f5e-48a0-ac46-2e762c4503d9" />

**Comment by msehabibur** [comment]
https://github.com/microsoft/mattergen/issues/46

**Comment by msehabibur** [comment]
<img width="1117" alt="Image" src="https://github.com/user-attachments/assets/c48822b1-7edf-4719-b695-ed205c3ecb10" />

**Comment by msehabibur** [comment]
@danielzuegner @txie-93

**Comment by msehabibur** [comment]
@Ruziy

**Comment by msehabibur** [comment]
Hi @danielzuegner this is the CLI command 

#!/bin/sh -l
#SBATCH -A standby
#SBATCH --nodes=1
#SBATCH --gpus-per-node=1
#SBATCH --mem=12G
#SBATCH --ntasks-per-node=1
#SBATCH -t 4:00:00
#SBATCH --job-name MatGL-PES
#SBATCH --constraint='a100'     #choose the specific GPUs
source .venv/bin/activate
mattergen-train data_module=chalcogenide ~trainer.logger
export RESULTS_PATH=results/ 
#mattergen-generate $RESULTS_PATH --model_path=/depot/amannodi/data/Habibur_Projects/mattergen/outputs/singlerun/2025-04-05/10-52-38/ --batch_size=16

**Comment by msehabibur** [comment]
@danielzuegner this is the config.yaml from the trained model

auto_resume: true
checkpoint_path: null
data_module:
  _recursive_: true
  _target_: mattergen.common.data.datamodule.CrystDataModule
  average_density: 0.05771451654022283
  batch_size:
    test: 16
    train: 128
    val: 16
  dataset_transforms:
  - _partial_: true
    _target_: mattergen.common.data.dataset_transform.filter_sparse_properties
  max_epochs: 2
  num_workers:
    test: 0
    train: 0
    val: 0
  properties: []
  root_dir: /depot/amannodi/data/Habibur_Projects/mattergen/mattergen/../datasets/cache/chalcogenide
  test_dataset:
    _target_: mattergen.common.data.dataset.CrystalDataset.from_cache_path
    cache_path: /depot/amannodi/data/Habibur_Projects/mattergen/mattergen/../datasets/cache/chalcogenide/test
    dataset_transforms:
    - _partial_: true
      _target_: mattergen.common.data.dataset_transform.filter_sparse_properties
    properties: []
    transforms:
    - _partial_: true
      _target_: mattergen.common.data.transform.symmetrize_lattice
    - _partial_: true
      _target_: mattergen.common.data.transform.set_chemical_system_string
  train_dataset:
    _target_: mattergen.common.data.dataset.CrystalDataset.from_cache_path
    cache_path: /depot/amannodi/data/Habibur_Projects/mattergen/mattergen/../datasets/cache/chalcogenide/train
    dataset_transforms:
    - _partial_: true
      _target_: mattergen.common.data.dataset_transform.filter_sparse_properties
    properties: []
    transforms:
    - _partial_: true
      _target_: mattergen.common.data.transform.symmetrize_lattice
    - _partial_: true
      _target_: mattergen.common.data.transform.set_chemical_system_string
  transforms:
  - _partial_: true
    _target_: mattergen.common.data.transform.symmetrize_lattice
  - _partial_: true
    _target_: mattergen.common.data.transform.set_chemical_system_string
  val_dataset:
    _target_: mattergen.common.data.dataset.CrystalDataset.from_cache_path
    cache_path: /depot/amannodi/data/Habibur_Projects/mattergen/mattergen/../datasets/cache/chalcogenide/val
    dataset_transforms:
    - _partial_: true
      _target_: mattergen.common.data.dataset_transform.filter_sparse_properties
    properties: []
    transforms:
    - _partial_: true
      _target_: mattergen.common.data.transform.symmetrize_lattice
    - _partial_: true
      _target_: mattergen.common.data.transform.set_chemical_system_string
lightning_module:
  _target_: mattergen.diffusion.lightning_module.DiffusionLightningModule
  diffusion_module:
    _target_: mattergen.diffusion.diffusion_module.DiffusionModule
    corruption:
      _target_: mattergen.diffusion.corruption.multi_corruption.MultiCorruption
      discrete_corruptions:
        atomic_numbers:
          _target_: mattergen.diffusion.corruption.d3pm_corruption.D3PMCorruption
          d3pm:
            _target_: mattergen.diffusion.d3pm.d3pm.MaskDiffusion
            dim: 101
            schedule:
              _target_: mattergen.diffusion.d3pm.d3pm.create_discrete_diffusion_schedule
              kind: standard
              num_steps: 1000
          offset: 1
      sdes:
        cell:
          _target_: mattergen.common.diffusion.corruption.LatticeVPSDE.from_vpsde_config
          vpsde_config:
            beta_max: 20
            beta_min: 0.1
            limit_density: 0.05771451654022283
            limit_var_scaling_constant: 0.25
        pos:
          _target_: mattergen.common.diffusion.corruption.NumAtomsVarianceAdjustedWrappedVESDE
          limit_info_key: num_atoms
          sigma_max: 5.0
          wrapping_boundary: 1.0
    loss_fn:
      _target_: mattergen.common.loss.MaterialsLoss
      d3pm_hybrid_lambda: 0.01
      include_atomic_numbers: true
      include_cell: true
      include_pos: true
      reduce: sum
      weights:
        atomic_numbers: 1.0
        cell: 1.0
        pos: 0.1
    model:
      _target_: mattergen.denoiser.GemNetTDenoiser
      atom_type_diffusion: mask
      denoise_atom_types: true
      gemnet:
        _target_: mattergen.common.gemnet.gemnet.GemNetT
        atom_embedding:
          _target_: mattergen.common.gemnet.layers.embedding_block.AtomEmbedding
          emb_size: 512
          with_mask_type: true
        cutoff: 7.0
        emb_size_atom: 512
        emb_size_edge: 512
        latent_dim: 512
        max_cell_images_per_dim: 5
        max_neighbors: 50
        num_blocks: 4
        num_targets: 1
        otf_graph: true
        regress_stress: true
        scale_file: /depot/amannodi/data/Habibur_Projects/mattergen/mattergen/common/gemnet/gemnet-dT.json
      hidden_dim: 512
      property_embeddings: {}
      property_embeddings_adapt: {}
    pre_corruption_fn:
      _target_: mattergen.property_embeddings.SetEmbeddingType
      dropout_fields_iid: false
      p_unconditional: 0.2
  optimizer_partial:
    _partial_: true
    _target_: torch.optim.Adam
    lr: 0.0001
  scheduler_partials:
  - frequency: 1
    interval: epoch
    monitor: loss_train
    scheduler:
      _partial_: true
      _target_: torch.optim.lr_scheduler.ReduceLROnPlateau
      factor: 0.6
      min_lr: 1.0e-06
      patience: 100
      verbose: true
    strict: true
load_original: false
params: {}
trainer:
  _target_: pytorch_lightning.Trainer
  accelerator: gpu
  accumulate_grad_batches: 1
  callbacks:
  - _target_: pytorch_lightning.callbacks.LearningRateMonitor
    log_momentum: false
    logging_interval: step
  - _target_: pytorch_lightning.callbacks.ModelCheckpoint
    every_n_epochs: 1
    filename: '{epoch}-{loss_val:.2f}'
    mode: min
    monitor: loss_val
    save_last: true
    save_top_k: 1
    verbose: false
  - _target_: pytorch_lightning.callbacks.TQDMProgressBar
    refresh_rate: 50
  - _target_: mattergen.common.data.callback.SetPropertyScalers
  check_val_every_n_epoch: 1
  devices: 1
  gradient_clip_algorithm: value
  gradient_clip_val: 0.5
  max_epochs: 2
  num_nodes: 1
  precision: 32
  strategy:
    _target_: pytorch_lightning.strategies.ddp.DDPStrategy
    find_unused_parameters: true

**Comment by danielzuegner** [comment]
Hi @msehabibur,

It looks like the model you've trained is unconditional:

```yaml
...
properties: []
...
property_embeddings: {}
property_embeddings_adapt: {}
...
```

Have you followed the steps from the [README](https://github.com/microsoft/mattergen?tab=readme-ov-file#fine-tune-on-your-own-property-data)?

**Comment by msehabibur** [comment]
@danielzuegner Yes, I edited the global.py to include dft_band_gap and also added the dft_band_gap.yaml 

<img width="1288" alt="Image" src="https://github.com/user-attachments/assets/0a30d920-73a5-4a98-815a-895c5b7e521a" />
<img width="1788" alt="Image" src="https://github.com/user-attachments/assets/7092817b-0d14-433f-8b61-bfed72793883" />

**Comment by danielzuegner** [comment]
Hi @msehabibur,

`dft_band_gap.yaml` already [exists in the repository](https://github.com/microsoft/mattergen/blob/main/mattergen/conf/lightning_module/diffusion_module/model/property_embeddings/dft_band_gap.yaml), and `dft_band_gap` is already included in [`globals.py`](https://github.com/microsoft/mattergen/blob/main/mattergen/common/utils/globals.py#L52). So It's unclear to me what changes you've made? 

Further, note that `mattergen-train data_module=chalcogenide ~trainer.logger` is a command which trains an unconditional model, see the description from the README:

> You can train the MatterGen base model on `mp_20` using the following command.

For conditional training you will need to add the property to the training, similar to what happens in the finetuning, e.g., `+lightning_module/diffusion_module/model/property_embeddings@adapter.adapter.property_embeddings_adapt.$PROPERTY=$PROPERTY ~trainer.logger data_module.properties=["$PROPERTY"]`.

**Comment by msehabibur** [comment]
I used this 

<img width="1038" alt="Image" src="https://github.com/user-attachments/assets/ceee96aa-237b-4963-9a48-6a34f88d0a25" />

<img width="1617" alt="Image" src="https://github.com/user-attachments/assets/0140c8bf-77a9-4143-8de5-41056a3cc3a5" />

**Comment by msehabibur** [comment]
Hi @danielzuegner I tried this 

#!/bin/sh -l
#SBATCH -A standby
#SBATCH --nodes=1
#SBATCH --gpus-per-node=1
#SBATCH --mem=12G
#SBATCH --ntasks-per-node=1
#SBATCH -t 4:00:00
#SBATCH --job-name MatGL-PES
#SBATCH --constraint='a100'

source .venv/bin/activate

export PROPERTY=dft_band_gap

mattergen-train \
  data_module=chalcogenide \
  +lightning_module.diffusion_module.model.property_embeddings@adapter.property_embeddings_adapt.${PROPERTY}=${PROPERTY} \
  ~trainer.logger \
  data_module.properties=["${PROPERTY}"]

export RESULTS_PATH=results/


But it shows 

MODELS_PROJECT_ROOT: /depot/amannodi/data/Habibur_Projects/mattergen/mattergen
In 'lightning_module/diffusion_module/model/mattergen': Could not find 'lightning_module/property_embeddings/dft_band_gap'

Config search path:
        provider=hydra, path=pkg://hydra.conf
        provider=main, path=file:///depot/amannodi/data/Habibur_Projects/mattergen/mattergen/conf
        provider=schema, path=structured://

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

**Comment by msehabibur** [comment]
<img width="617" alt="Image" src="https://github.com/user-attachments/assets/070dd680-85a0-4dbd-bb29-72215357ffca" />
<img width="1141" alt="Image" src="https://github.com/user-attachments/assets/950985a6-6671-4f4b-b932-cab2fe67d2d6" />
<img width="1212" alt="Image" src="https://github.com/user-attachments/assets/abf326d4-5370-4c63-9023-7a7a411dcfbf" />
<img width="1623" alt="Image" src="https://github.com/user-attachments/assets/99d35ed4-029b-4b72-ae96-c7d406c5f9ab" />
<img width="1618" alt="Image" src="https://github.com/user-attachments/assets/039125cf-13dc-4ecf-b7f0-09c697070249" />
<img width="1422" alt="Image" src="https://github.com/user-attachments/assets/491d40ac-af5d-47f8-8f2e-0bdb73bc20b6" />

**Comment by msehabibur** [comment]
@danielzuegner any feedback? Thanks in advance!

**Comment by danielzuegner** [comment]
Hi @msehabibur,

looks like you're training from scratch rather than fine-tune, let's try this config override instead: `+lightning_module.diffusion_module.model.property_embeddings@lightning_module.diffusion_module.model.property_embeddings.${PROPERTY}=${PROPERTY}`. Let me know if that works.

**Comment by msehabibur** [comment]
Hi @danielzuegner it shows 

bash run.sh 
MODELS_PROJECT_ROOT: /depot/amannodi/data/Habibur_Projects/mattergen/mattergen
In 'lightning_module/diffusion_module/model/mattergen': Could not find 'lightning_module/property_embeddings/dft_band_gap'

Config search path:
        provider=hydra, path=pkg://hydra.conf
        provider=main, path=file:///depot/amannodi/data/Habibur_Projects/mattergen/mattergen/conf
        provider=schema, path=structured://

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

**Comment by msehabibur** [comment]
bash run.sh 
MODELS_PROJECT_ROOT: /depot/amannodi/data/Habibur_Projects/mattergen/mattergen
Override +lightning_module.property_embeddings@lightning_module.property_embeddings.dft_band_gap=dft_band_gap looks like a config group override, but config group 'lightning_module.property_embeddings' does not exist.

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

**Comment by msehabibur** [comment]
<img width="1323" alt="Image" src="https://github.com/user-attachments/assets/52dee5fa-604d-4eb0-b8d9-a2e2bed61ceb" />

**Comment by danielzuegner** [comment]
I see, can you try this instead? `lightning_module/diffusion_module/model/property_embeddings@lightning_module.diffusion_module.model.property_embeddings.${PROPERTY}=${PROPERTY}`

**Comment by msehabibur** [comment]
#!/bin/sh -l
#SBATCH -A standby
#SBATCH --nodes=1
#SBATCH --gpus-per-node=1
#SBATCH --mem=12G
#SBATCH --ntasks-per-node=1
#SBATCH -t 4:00:00
#SBATCH --job-name MatGL-PES
#SBATCH --constraint='a100'

source .venv/bin/activate

export PROPERTY=dft_band_gap

mattergen-train \
  data_module=chalcogenide \
  lightning_module/diffusion_module/model/property_embeddings@lightning_module.diffusion_module.model.property_embeddings.${PROPERTY}=${PROPERTY} \
  ~trainer.logger \
  data_module.properties=["${PROPERTY}"]

export RESULTS_PATH=results/

**Comment by msehabibur** [comment]
bash run.sh 
MODELS_PROJECT_ROOT: /depot/amannodi/data/Habibur_Projects/mattergen/mattergen
Could not override 'lightning_module/diffusion_module/model/property_embeddings@lightning_module.diffusion_module.model.property_embeddings.dft_band_gap'. No match in the defaults list.
To append to your default list use +lightning_module/diffusion_module/model/property_embeddings@lightning_module.diffusion_module.model.property_embeddings.dft_band_gap=dft_band_gap

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

**Comment by danielzuegner** [comment]
Sorry, I forgot the leading plus sign in my last message. Please try `+lightning_module/diffusion_module/model/property_embeddings@lightning_module.diffusion_module.model.property_embeddings.${PROPERTY}=${PROPERTY}`.

**Comment by msehabibur** [comment]
@danielzuegner Thank you so much! I solved this issue. Can I ask, what is the CLI to restart the training of a model from prior checkpoint?

**Comment by danielzuegner** [comment]
Hi @msehabibur,

you'll have to provide the override `hydra.run.dir=/path/to/your-existing-run` and then it should pick up the existing checkpoint.

Daniel
